File: project/src/main/java/edu/tum/ase/project/model/SourceFile.java
package edu.tum.ase.project.model;

import org.hibernate.annotations.GenericGenerator;

import com.fasterxml.jackson.annotation.JsonBackReference;

import jakarta.persistence.*;

@Entity
@Table(name = "source_files")
public class SourceFile {
    @Id
    @GeneratedValue(generator = "system-uuid")
    @GenericGenerator(name = "system-uuid", strategy = "uuid")
    @Column(name = "source_file_id")
    private String id;

    @Column(name = "file_name", nullable = false, unique = true)
    private String fileName;

    @Column(name = "code", nullable = false)
    private String code;

    @ManyToOne
    @JsonBackReference
    @JoinColumn(name = "project", nullable = false)
    private Project project;

    protected SourceFile() {
        // no-args constructor required by JPA spec
        // this one is protected since it shouldn't be used directly
    }

    public SourceFile(String fileName, String code, Project project) {
        this.fileName = fileName;
        this.code = code;
        this.project = project;
    }

    // getters and setters
    public String getId() {
        return id;
    }

    public String getFileName() {
        return fileName;
    }

    public String getCode() {
        return code;
    }

    public Project getProject() {
        return project;
    }

    public void setId(String id) {
        this.id = id;
    }

    public void setFileName(String fileName) {
        this.fileName = fileName;
    }

    public void setCode(String code) {
        this.code = code;
    }

    public void setProject(Project project) {
        this.project = project;
    }
}

File: ui/frontend/src/app/classes/sourceFile/sourceFile.ts
export class SourceFile {
  id: number;
  fileName: string;
  code: string;
  project: string;
  stdout: string = '';
  stderr: string = '';
  compilable: boolean = false;
  constructor(
    id: number,
    fileName: string,
    code: string,
    project: string,
  ) {
    this.id = id;
    this.fileName = fileName;
    this.code = code;
    this.project = project;
  }
}

export class CompiledSourceFileDTO {
  fileName: string;
  code: string;
  stdout: string;
  stderr: string;
  compilable: boolean;
  constructor(fileName: string, code: string, stdout: string, stderr: string, compilable: boolean) {
    this.fileName = fileName;
    this.code = code;
    this.stdout = stdout;
    this.stderr = stderr;
    this.compilable = compilable;
  }
}


export class CreateSourceFileDTO {
  fileName: string;
  code: string;
  project: string;
  constructor(fileName: string, code: string, project: string) {
    this.fileName = fileName;
    this.code = code;
    this.project = project;
  }
}

export class SourceFileDTO {
  fileName: string;
  code: string;
  constructor(fileName: string, code: string) {
    this.fileName = fileName;
    this.code = code;
  }
}
